### PR TITLE
Replace dynamic client with metadata client in tenantcontroller

### PR DESCRIPTION
**Pull request by Sindica**
First commit is a squash from https://github.com/CentaurusInfra/arktos/pull/935/commits
2nd & 3rd commits were cherry-picked from 1.18 - use metadata informer instead of dynamic informer in KCM
The rest are Arktos changes to replace dynamic client by metadata client in tenant controller, plus UTs and integration tests.

Reason:
This is a continuation of https://github.com/CentaurusInfra/arktos/pull/935. This change is focus on replace dynamic client with metadata client in controller manager. It completes the purpose of 935.

### CODE REVIEW

File: test/integration/tenantcontroller/tenantcontroller_test.go [hunk 1]
```diff
@@ -50,7 +50,7 @@ func setup(t *testing.T) (*httptest.Server, framework.CloseFunc, *tenantcontroll
 	resyncPeriod := 12 * time.Hour
 	informerSet := informers.NewSharedInformerFactory(clientset.NewForConfigOrDie(restclient.AddUserAgent(configs, "cronjob-informers")), resyncPeriod)
 
-	dynamicClient, err := dynamic.NewForConfig(configs)
+	metadataClient, err := metadata.NewForConfig(configs)
 	if err != nil {
 		t.Fatalf("Errror creating dynamic client")
```
Comment (chenqianfzh) [inline test/integration/tenantcontroller/tenantcontroller_test.go:55]: Nit: update the Fatalf message to be about metadata client.

File: test/integration/garbagecollector/garbage_collector_test.go [hunk 2]
```diff
@@ -242,15 +242,15 @@ func setupWithServer(t *testing.T, result *kubeapiservertesting.TestServer, work
 		t.Fatalf("failed to create dynamicClient: %v", err)
```
Comment (chenqianfzh) [inline test/integration/garbagecollector/garbage_collector_test.go:242]: Why we need both metadataclient and dynamic client? I am curious about it.

Comment (Sindica) [reply test/integration/garbagecollector/garbage_collector_test.go:242]: Looks like community hasn't migrated their integration test from dynamic client to metadata client. It is probably intended for backward compatibility. As long as KCM is fully free of dynamic client, I am ok having dynamic client in integration tests.

Comment (Sindica) [reply test/integration/garbagecollector/garbage_collector_test.go:242]: https://github.com/kubernetes/kubernetes/blob/master/test/integration/garbagecollector/garbage_collector_test.go

**Review by chenqianfzh — APPROVED** [review]

**Pull request merged**